src/engine/app.rs

```diff
@@ -221,51 +221,57 @@ impl Application {
 }
 
 /// Helper builder for Applications.
-pub struct ApplicationBuilder<T>
+pub struct ApplicationBuilder<'a, T>
     where T: State + 'static
 {
     config: DisplayConfig,
     initial_state: T,
-    planner: Planner<()>,
+    dispatcher_builder: DispatcherBuilder<'a, 'a>,
+    world: World,
 }
 
-impl<T> ApplicationBuilder<T>
+impl<'a, T> ApplicationBuilder<'a, T>
     where T: State + 'static
 {
     /// Creates a new ApplicationBuilder with the given initial game state and
     /// display configuration.
-    pub fn new(initial_state: T, cfg: DisplayConfig) -> ApplicationBuilder<T> {
-        let pool = Arc::new(ThreadPool::new(num_cpus::get()));
+    pub fn new(initial_state: T, cfg: DisplayConfig) -> ApplicationBuilder<'a, T> {
+        let pool = Arc::new(ThreadPool::new( ::rayon::Configuration::new().num_threads(num_cpus::get())).unwrap());
 
         ApplicationBuilder {
             config: cfg,
             initial_state: initial_state,
-            planner: Planner::from_pool(World::new(), pool),
+            dispatcher_builder: DispatcherBuilder::new().with_pool(pool),
+            world: World::new(),
         }
     }
 
     /// Registers a given component type.
-    pub fn register<C>(mut self) -> ApplicationBuilder<T>
+    pub fn register<C>(mut self) -> ApplicationBuilder<'a, T>
         where C: Component
     {
-        {
-            let world = &mut self.planner.mut_world();
-            world.register::<C>();
-        }
+        self.world.register::<C>();
+        self
+    }
+
+    /// Adds a given system `pro`, assigns it the string identifier `name`,
+    /// and marks it with the runtime priority `pri`.
+    pub fn add_barrier(mut self) -> ApplicationBuilder<'a, T> {
+        self.dispatcher_builder = self.dispatcher_builder.add_barrier();
         self
     }
 
     /// Adds a given system `pro`, assigns it the string identifier `name`,
     /// and marks it with the runtime priority `pri`.
-    pub fn with<S>(mut self, sys: S, name: &str, pri: Priority) -> ApplicationBuilder<T>
-        where S: System<()> + 'static
+    pub fn with<S>(mut self, sys: S, name: &str, dep: &[&str]) -> ApplicationBuilder<'a, T>
+        where for<'b> S: System<'b> + Send + 'a
     {
-        self.planner.add_system::<S>(sys, name, pri);
+        self.dispatcher_builder = self.dispatcher_builder.add(sys, name, dep);
```

Comment:
Not sure if we should just wrap the functionality. Could we just let the user pass in the dispatcher with all systems already added?